LwM2M data formatting bugfixes

mike-scott opened the pull request:
This patchset looks quite long for a series of bugfixes.  However, I've tried to address several bugs exposed in our data formatting layer knowing that more data formatters are going to be added for LwM2M 1.1 support (CBOR).   Each patch should be atomic and easily consumable for the reviewer.  If not let me know.

Addresses the following issues:
- Having 1 main read_op function which processed all of the data formatters was prone to breakage, so it needed to be split up.
- TLV formatter didn't indicate "object instance" boundry when more than 1 object was being returned.
- JSON formatter had all sorts of problems and only the READ related issues are addressed in this series
- Plain text formatter should only return data for a single resource request

The initial goal of this series was to fix https://github.com/zephyrproject-rtos/zephyr/issues/9470

mike-scott committed:
net: lwm2m: simplify MATCH_ logic in do_read_op()

Remove over-complicated match_type logic in do_read_op().  Replace
MATCH_* checks with actual path->level values.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: fix logic for lwm2m_next_engine_obj_inst()

The object instance list isn't sorted by object instance id.  Let's
simplify this and fix the logic in lwm2m_next_engine_obj_inst() to make
sure that we always get the NEXT object instance by value of
obj_inst_id, not just the next object instance in the list.

NOTE: This change removes the "last" object instance pointer from the
parameters of lwm2m_next_engine_obj_inst().  Some of the logic to return
a NULL value for the end of the list has to be moved back into
do_read_op().

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: fix reading multiple objects that don't start at 0

Let's correct the starting logic in do_read_op() to not assume
a default value of 0 will be present for the first object, when
reading multiple objects.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: refactor put_begin_ri/put_end_ri into generic functions

In order to re-use the put_begin_ri / put_end_ri logic, let's create
generic functions for them: put_begin_tlv and put_end_tlv

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: in oma_tlv_put don't re-add value when insert is true

We set "insert" to true when the value is already in the buffer, but
we need to insert a TLV to denote things like RESOURCE_INSTANCE or
OBJECT_INSTANCE.  In this case, let's not re-add the value.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott commented:
@nashif I know this is a rather long patchset to fix bugs and the 1.13 deadline is very very close.  If you don't feel that this should go into 1.13, let me know.

@therealprof feel free to review.  These address several issues which are probably affecting you.

mike-scott commented:
Rebased and fixed the SHIPPABLE commit desc complaint

therealprof reviewed (approved):
Other than one comment I gave it a spin on a Nucleo and so far it works great.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_rw_plain_text.c:
```diff
@@ -334,6 +333,64 @@ const struct lwm2m_reader plain_text_reader = {
+	int ret = 0, index;
+
+	/* Plain text can only return single resource */
+	if (path->level >= 3) {
```
How about this instead here:
```
if (path->level == 3) {
```

therealprof replied:
I would assume that this is correct but I've no idea what exactly `level` is expressing.

mike-scott replied:
"level" is the depth of the uri.  Since we're using unsigned integer types and a value of "0" is valid, we need a way of knowing how much of the URI was set.

IE:
/3 == level 1, with values: obj_id=3, obj_inst_id=0, res_id=0
/3/0 == level 2, with values: obj_id=3, obj_inst_id=0, res_id=0
/3/0/0 == level 3, with values: obj_id=3, obj_inst_id=0, res_id=0

therealprof replied:
That matches with my assumption so `== 3` would be correct. The only thing I'm not clear about is whether a level == 4 might be used to express multiple resources. If so then `>= 3` would be wrong because multiple resources can not be expressed in plain text. If level == 4 cannot happen anyways then then point is moot. ;)

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -681,18 +681,20 @@ lwm2m_get_engine_obj_inst(int obj_id, int obj_inst_id)
 }
```
Nice catch.  Fixing.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_rw_oma_tlv.c:
```diff
@@ -717,6 +718,122 @@ const struct lwm2m_reader oma_tlv_reader = {
 	get_opaque
 };
 
+int do_read_op_tlv(struct lwm2m_engine_obj *obj,
```
Yeah, at the end of all the work, they do seem very similar.  The outlier is plain-text where you have almost no processing at all since it only handles single results.

Let me see what I can do to cut down on duplicate code there.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_object.h:
```diff
@@ -271,6 +271,9 @@ struct lwm2m_output_context {
 	u8_t writer_flags;
+
+	/* private output data */
+	void *data;
```
Yep will change

mike-scott commented:
I'm going to fix the duplicated code by leaving the majority of the do_read_op() code in lwm2m_engine.c but under a different name (adding the new version of do_read_op).

Each formatter can do some initial setup or reject invalid requests and then call back into that shared function.  This also means I will keep the put_begin / put_end and put_begin_ri APIs in the output_formatter structure, as well as add:
put_begin_oi / put_end_oi (object instance)
put_begin_r / put_end_r (resource)

That way formatters can perform processing at almost any point.

mike-scott committed:
net: lwm2m: allow formatters to perform processing prior to read_op

Data formatters are becoming too complex for a simple do_read_op()
function to handle all in one place.  Also, more data formatters are
going to be added for LwM2M v1.1 support in the future.

In order for data formatters to perform internal setup or deny
invalid requests (specific to the formatter's logic), let's
establish do_read_op_* functions in each formatter.

Once the internal processing is done, they can call back into the
more generic lwm2m_perform_read_op function.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: plain-text: process only reads for a specific resource

The plain-text format only supports READ op for a specific resource.
In all other cases return NOT_ALLOWED.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: introduce output context user_data

Data formatters have various private state variables which are
currently located in the output context structure.  Let's add
a place where data formatters can store a pointer to their
private data so that as we add more formatters the output
context doesn't get cluttered up.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: json formatter use private data

Use newly introduced private data pointer in output context to
store JSON formatter information.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: tlv formatter use private data

Use newly introduced private data pointer in output context to
store TLV formatter information.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: remove unused members from lwm2m_output_context

Now that formatters use their own private data to hold state,
let's remove the old member variables from lwm2m_output_context
which are now unused.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: optimize variable order in lwm2m_perform_read_op()

Let's sort by largest to smallest so that we don't leave odd gaps
in memory.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: correct placement of put_begin/put_end in READ op

The put_begin / put_end calls are to be used at the very beginning
and end of processing a READ op.  Let's correct that logic.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: fix JSON format for multi-instance reads

When reading multiple instances, the base name value should not
include an object instance id.  The object instance id is added
to the individual resource name values.

Accomplish this by saving the original path level and adjusting
the (base) name where needed.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: optimize lwm2m_perform_read_op()

Optimize the resource processing loop to avoid extra
assignments before checking if we need to process the
actual resource.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: introduce put_begin/end for object instance and resources

Data formatters may need to process data at the beginning and end of
each object instance and/or resource.  Currently, they can only add
processing at the beginning and end of resource instances.

Let's establish put_begin/end_oi (object instance) and put_begin/end_r
(resource) API functions that data formatters can use for this purpose.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: store a backup of the entire path in perform_read_op

Currently, we only save the resource id of the incoming path setting.
In the future, we will need to change other values in order to process
multi-instance READ operations.

Let's save and restore the entire path only at the beginning and end
of processing.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: implement begin/end processing for obj inst and resources

Implement put_begin/end calls for object instance and resource
processing in lwm2m_perform_read_op()

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: TLV: mark object instance boundry when needed

Let's implement put_begin/end_oi functions in the TLV formatter
so to mark the boundry of an object instance when more than 1
object instance is returned.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/9470

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott commented:
Addressed comments and implemented the changes described above.  The diff is rather large from the previous version, but here it is: https://hastebin.com/azawokalax.diff

@therealprof mind giving this version a spin?

mike-scott commented:
Small bugfix and SHIPPABLE complaint

therealprof reviewed (approved):
Tested on Nucleo-F746, works fine.

jukkar commented on subsys/net/lib/lwm2m/lwm2m_rw_json.c:
```diff
@@ -453,6 +508,10 @@ static size_t put_bool(struct lwm2m_output_context *out,
 const struct lwm2m_writer json_writer = {
 	put_begin,
```
Hmm, I missed this one earlier. Doing it like this is error prone, please change this to use the alternate syntax
```
.put_begin = put_begin,
.put_end = put_end,
...
```
This way the ordering of the fields does not matter and we can safely add new fields into the struct if needed.

jukkar commented on subsys/net/lib/lwm2m/lwm2m_rw_oma_tlv.c:
```diff
@@ -692,6 +785,10 @@ static size_t get_opaque(struct lwm2m_input_context *in,
 }
 
 const struct lwm2m_writer oma_tlv_writer = {
+	NULL,
```
same thing here about the initialization

jukkar commented on subsys/net/lib/lwm2m/lwm2m_rw_plain_text.c:
```diff
@@ -309,6 +310,10 @@ static size_t get_opaque(struct lwm2m_input_context *in,
 }
 
 const struct lwm2m_writer plain_text_writer = {
+	NULL,
```
ditto

jukkar reviewed (dismissed):
Almost there but couple of minor things still found.

mike-scott committed:
net: lwm2m: fix formatter reader/writer initialization syntax

For ease of maintenance, let's swap the reader/writer initialization
syntax to:
.put_begin = put_begin,
.put_end = put_end,
...

This way we only assign used fields and adding new ones later is
less error prone.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott commented:
Added a new patch to the end of the series cleaning up the reader / writer initialization syntax (per @jukkar ).  If that's not acceptable, I can look at blending those changes back into the patch set, but IMHO it's much clearer to see this change separately.

mike-scott commented:
While working on this series, I also noticed that JSON WRITE support was never implemented.  I worked up a patch series, and I'll submit that for 1.14 after I've done more testing.
